### ISSUE TITLE
Provide TypeScript interface while extending `Model` class

### ISSUE THREAD

**Issue by cwadrupldijjit** [Vincit/objection.js#2237]
- [x] Objection.js version:  3.0.1
- [x] Type of update:  Feature

This affects TypeScript users.

My company prefers to use interfaces to type things over classes if we don't expect to run `new` on them.  For instance, if we get a request from the API with the shape of a model to create or update, we have an interface for it.  As such, we would be using an interface for the `req.body` and then feeding that into Objection to create/update an item.

Currently, we have model classes that are created and only specify the static getters (such as `tableName` and `jsonSchema`) and a separate interface that contains properties and expected values for them.  This also results in no record information from `.query` calls on the class itself.  There is also currently no way to provide an interface and have that populate the fields you will get back from the record.

I've seen some extension of query building classes that can do something similar (as far as I can see) and another option that doesn't change Objection itself would be to specify the properties (again) (and possibly through an `implements <interface-name>`, for type-safety).

But the solution I'd prefer follows a path like:
```ts
class User extends Model<UserInit> {
    // no need to redefine "UserInit" properties--they come from the generic type
}
```

I tried modifying the typing file itself in my local `node_modules` folder because I wanted to see what it would take to resolve it.  I came across many things that don't work and one that works okay for my needs.

First, I want to explain one way that doesn't work:

```ts
class Model<PropertyType extends object = object> extends PropertyType { // generic ignored; as good as extending `object`...
```

The result that works the best uses `type` initialization since it's blending things (by necessity) by using unions between the types.

The changes look like this (to the `typings/objection/index.d.ts`):

```ts
export class _Model { // renamed from "Model"
...
}
export type Model<PropertyType extends object = object> = _Model & PropertyType;
export const Model: typeof _Model & (new <PropertyType extends object = object>() => Model<PropertyType>;
```

The biggest drawback of this I can see is that the generic (even with the auto-assignment) doesn't like to not be filled with something.  As such, every usage of `Model` requires a generic to be included with it:

```ts
class User extends Model<UserInit> {} // Works

class UserProfile extends Model {} // Breaks
```

Is there any chance we could get something like the change I made into the package?  The biggest problem is that it's a breaking change, but I'm not sure what I can do about it.  This discussion could include some of the notes I made during the course of sussing out a better solution.

**Comment by cwadrupldijjit** [comment]
I might bump here, but if anybody would like to have the workaround I put into the project I'm working on, I made the same changes as above, but I moved them into one of my own files.  Instead of using `Model` directly, I'm assigning it to a different variable (in this case, `Db`) and setting that type instead.

Meaning, the final code that I export (and use everywhere else in the application instead of the `Model` class directly) looks like:
```ts
import { Model } from 'objection';

// register the knex instance, etc.

export type Db<PropertyType extends object = object> = Model & PropertyType;
export const Db: typeof Model & (new <PropertyType extends object = object>() => Db<PropertyType>) = Model as any; // type cast here because it'll complain otherwise
```

Still would be better to have that capability baked in, but whatever works, I suppose.

**Comment by damusix** [comment]
@cwadrupldijjit Agreed 100%

I am personally taking this approach:


```ts

type UserQbType<T extends Model> = (
    QueryBuilder<T> & {
        someFn(): QueryBuilder<T>
    }
)

export interface User extends TimestampableInterface {

    id: string;
    email: string;
    confirmed_at: Nullable<Date>;
    tags?: Tag[]

    QueryBuilderType: UserQbType<this>;

}

export class User extends Model {

    static QueryBuilder = MyUserQueryBuilder;

    id!: string;
    confirmed_at!: Nullable<Date>;
    invited_by_id!: Nullable<string>;

    static tableName = 'users';

    static relationMappings = {


        tags: {
            modelClass: __dirname + '/Tag',
            relation: Model.ManyToManyRelation,
            join: {
                from: 'users.id',
                through: {
                    from: 'user_tags.user_id',
                    to: 'user_tags.tag_id',
                },
                to: 'tags.id',
            }
        }
    };
}

```

Works well and merges `interface User {}` with `class User {}` and you can extend the interface with multiple other interfaces.

One particular thing that's very annoying is the fact that [even though the docs tell you to extend the class for query builders](https://vincit.github.io/objection.js/recipes/custom-query-builder.html#extending-the-query-builder-in-typescript), you can't extend the class because it expects it to be instantiated, and your TS configs may give you shit for using  `!`. If you define it on the interface however, it still will pick it up and merge with the class. It's less hacky that needing to do some weird generic that extends some kind of `AnyConstructor`

**Comment by cwadrupldijjit** [comment]
This seems interesting, @damusix.  Does the class automatically get updated along with the same-named interface when the interface is adjusted or does the class still need to be updated along with the interface?  At this point, I'd want both to be updated at the same time, and if this solution does that, then I'm willing to try that out for this project.

**Comment by damusix** [comment]
@cwadrupldijjit Yes it's going to assume anything you add to either side exists. Class is the implementation, interface is the typings. In fact that second of declarations on the class is useless for typings given that the values will be filled by the query on instantiation, but that's all objection's implementation.

```ts
class ... {
    // useless
    id!: string;
    confirmed_at!: Nullable<Date>;
    invited_by_id!: Nullable<string>;
}
```

**Comment by triv14** [comment]
Hi when i used these method, for async function that will return a Promise< User >, TS give me error for Type 'Model[]' is not assignable to type User[] cause it missing those fields in Model. Can you give me some pointers on how to take care of this error?

**Comment by Vectarian-Jorge** [comment]
I'm using `readonly` so that other people can't accidentally edit them as if that were to actually modify the values in the database. But it feels rather hacky to this to have the typings.
```
class PersonModel extends Model {
	static override get tableName() {
		return "public.person";
	}

	static override get idColumn() {
		return "id";
	}

	static override get jsonSchema() {
		return {
			type: "object",

			properties: {
				id: { type: "string" },
				foo: { type: "string" },
				...
			},
		};
	}

	static override get relationMappings() {
		return {
			pets: {
				relation: Model.HasManyRelation,
				modelClass: PetModel,
				join: {
					from: "public.person.id",
					to: "public.pet.owner",
				},
			},
			...
		};
	}

	readonly id!: string;
	readonly foo!: string;
	...

	readonly pets?: PetModel[];
}```
